### PR TITLE
Add constrained/flow layout to Cover block.

### LINKED ISSUE #43140: Add Layout support to Cover block
## What problem does this address?

In block themes, where each container block must specify its content width, Cover block has no settings to do so. This causes issues such as #37036.

An initial attempt to add Layout support to Cover (#37662) ran into trouble because of the complex markup structure of the block. The problem is similar to the Navigation block, which was addressed with a custom solution in #37473.

Additionally, the Cover block has a unique "Change content position" control in its toolbar, that allows aligning the content on a two dimensional axis that uses flexbox under the hood. If we migrate Cover to use Layout, it makes sense to replace this with the existing `flex` layout type.

Screen Shot 2022-08-11 at 5 43 53 pm

## What is your proposed solution?

We _could_ use the same custom solution as for Navigation block. That would work well for the flex layout scenario, but we'd also like Cover to work with the default (or content width enabled) layout, and that might be problematic because the max-width values we need for content width can't be hardcoded into the block stylesheet.

(As an aside, we'll need to dust off the [Layout type switcher](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/hooks/layout.js#L167) that currently isn't in use anywhere). Design might be needed because we haven't looked at it in a while 😄 

Ideally, we'd have a way to add the layout classes to the inner block wrapper (as [mentioned](https://github.com/WordPress/gutenberg/pull/37662#issuecomment-1003354817) on the original PR), so the layout styles would always work flawlessly, independent of block markup structure.

If #42485 goes forward, we could leverage a brand new Walker to do this, or we could modify [the existing logic](https://github.com/WordPress/gutenberg/blob/trunk/lib/block-supports/layout.php#L317-L318) slightly, but either way it looks like we'll need some classname or other way to identify the inner block wrapper. Should we use a classname, do we want to output it in the front end? Or just use it as a temporary internal hook to identify inner block containers? What other uses could such a thing have? (I'm thinking it should be added to all inner block containers, so we can be consistent with the layout logic.)

Any further ideas and feedback welcome!

### PR DESCRIPTION
**Pull request by tellthemachines**
## Why?

Closes #43140.

Adds support for constrained (as default) and flow layout to Cover block.

## How?

Now that #44600 is merged, layout classnames are applied to the block inner wrapper by default. This means that layout support will work in Cover block in the same way as it does in e.g. Group (but without the flex variations for now).

## Testing Instructions

1. Add a Cover block to a post;
2. Check that the Layout section now displays in the sidebar;
3. Check that constrained layout applies by default to Cover children;
4. Check that Cover children now have wide/full alignment controls where applicable.

TODO:

Existing Cover blocks should not change their layouts. I haven't done any testing for this yet, but anticipate some tweaks may be needed. I don't think a deprecation should be required because we're only adding functionality, not changing anything existing.

## Screenshots or screencast 
Screen Shot 2022-10-27 at 11 14 12 am

### CODE REVIEW

**Review by andrewserong — COMMENTED** [review]
Nice one, thanks for the quick follow-up!

> Existing Cover blocks should not change their layouts. I haven't done any testing for this yet, but anticipate some tweaks may be needed.

Yes, it looks like we'll need some kind of intervention to work out how existing blocks should be handled. From testing with a Cover block on trunk set to the Center Center justification in the toolbar controls, there's a change in how the blocks are laid out. Here's a before and after:

| Trunk | The same post loaded with this branch |
| --- | --- |
| image | image |

Also, it looks like the justification controls are exposed in the layout controls in the sidebar, which are different and conflict slightly with the ones the Cover block provides in the toolbar. E.g. justify left and right only work if the Toolbar control is set to Center Center:

image

Should the justification controls be hidden for the moment, or alternately, should they play nicely with the Cover block's ad hoc controls I wonder 🤔

File: packages/block-library/src/cover/edit/index.js
Comment (youknowriad) [inline packages/block-library/src/cover/edit/index.js]: Is there a way to inject this automatically in the block support instead? It seems this should be done automatically for blocks that have the "layout" block support enabled no?

Comment (tellthemachines) [reply packages/block-library/src/cover/edit/index.js]: It should, and this is one of the things we need to tackle before stabilising layout. Currently, the alignment controls of the child blocks depend on the layout settings being passed explicitly through innerBlocks. We have an open issue for it: #43731

I thought I'd go ahead with adding layout to Cover as it's one of the back-compat features requested by folks adding theme.json to their classic themes (so that Cover children can use the theme content width - see #33374 for more context). Once we have worked out a way to do this automatically, we can go back and remove the logic from existing blocks; we already pass layout to innerBlocks in most blocks that use it.

**Comment by tellthemachines** [comment]
Thanks for reviewing and testing, folks!

> From testing with a Cover block on trunk set to the Center Center justification in the toolbar controls, there's a change in how the blocks are laid out.

I think that's an inevitable consequence of making "constrained" the default, so I've made it default to "flow" for now. This won't be great for the classic theme folks wanting to experiment with theme.json though - the idea of having "constrained" as default was to make that transition process more seamless.

> Should the justification controls be hidden for the moment, or alternately, should they play nicely with the Cover block's ad hoc controls I wonder 🤔

Unfortunately Cover block matrix and the layout justification controls have completely different implementations, with the matrix styles coming from the block library stylesheet, so there's no way of making them work together unless we want to try re-implementing matrix as a layout-type control 😅 so I've hidden the justification controls for now.

**Comment by tellthemachines** [comment]
Thought I'd try what we're doing with the Group block and add constrained layout as a default block variation. This should mean that existing blocks will get flow layout applied but any new blocks added will be constrained. This seems like a good intermediate solution, and it's the same we adopted for Group. What does everyone think?

File: packages/block-library/src/cover/block.json
Comment (ntsekouras) [inline packages/block-library/src/cover/block.json:114]: I don't think we should add a new layout property that is only used in `constrained` layout..

File: packages/block-library/src/cover/variations.js
Comment (ntsekouras) [inline packages/block-library/src/cover/variations.js]: I don't think we need `isActive` and `scope:block|transform` here, since we don't have any other variations..

**Review by ntsekouras — COMMENTED** [review]
Thanks for your work @tellthemachines!

I'd feel more comfortable if this could get more testing before landing today and therefore in `6.2`. I haven't tested thoroughly but there are visible style changes in existing Cover blocks.

I have an example screenshot below that shows the margins are different for innerBlocks.

https://user-images.githubusercontent.com/16275880/216005660-aeeb0690-c529-45c8-a6d7-77c0763c84e7.mov

**Comment by tellthemachines** [comment]
Thanks for taking the time to review and test this @ntsekouras ! I thought with the latest commit existing blocks weren't changing anymore, but something must have escaped me. If you have a moment, could you share the markup for those test blocks where you see changes?

File: packages/block-library/src/cover/block.json
Comment (tellthemachines) [reply packages/block-library/src/cover/block.json:114]: It's a bit specific, but we do have a precedent in [allowOrientation](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/layouts/flex.js#L69) which is a valid property that only applies to flex layout (I don't think it's being used in any core blocks right now though). Do you see a better way of selectively disabling controls for a certain layout type?

Comment (ntsekouras) [reply packages/block-library/src/cover/block.json:114]: I'm not sure right now, but I guess it would need a new API maybe something like `{ [layout]: allowedControls }`? It would need exploration for sure to find what works and makes sense.

**Comment by ntsekouras** [comment]
>Thanks for taking the time to review and test this @ntsekouras ! I thought with the latest commit existing blocks weren't changing anymore, but something must have escaped me. If you have a moment, could you share the markup for those test blocks where you see changes?

What I did was just create a Cover block(copy many times 😄 ) and then set different content position to each. Then test that content with your PR.

Screenshot 2023-02-02 at 9 25 52 AM

**Comment by tellthemachines** [comment]
Oh, I think I see what you mean @ntsekouras. Currently, Cover block children are inheriting user agent margin styles unless they have any block- or global-level margins set. With the addition of layout, they instead get the layout margins applied, which means the first child has no margin and its siblings have a margin-top of whatever spacing value is set by the theme (or by core if the theme doesn't set any).

I'd be inclined to consider this an improvement instead of a regression 😅 because, although technically there is a change, that change makes spacing inside Cover consistent with the rest of the site, e.g. post content already has those spacing rules applied.

**Review by andrewserong — COMMENTED** [review]
> I'd be inclined to consider this an improvement instead of a regression 

This is a really interesting point, I think I'm inclined to agree that by using the layout support's `margin-block-start: 0` rules this is effectively a bug fix in that the positioning controls of the Cover block now work more consistently. In a test post where I tried out each of the permutations, I found it noticeable that in the center positions, with this PR applied, elements are correctly centred where they weren't before due to the paragraph's user agent margin styles, as you mention. E.g. here's a before / after of vertically centered, align right — note that in the before version, the browser default top margin of the paragraph block causes the blocks to sit too far down:

image

Similarly, it's quite noticeable when looking at top alignment:

image

Whereas, using the same pattern, the issue isn't noticeable in the bottom alignment, because the Buttons block is the lowest block visually, so doesn't have a browser default margin (it would be noticeable if the paragraph block were the lowest visually):

image

So, I think there's a compelling argument that by introducing the layout support to this block, it either fixes a bug in the behaviour of the alignment matrix, or is an enhancement to the alignment matrix, in that the positioning is now much more consistent in what it does — and likely closer to a user's expectations for how it would work.

The main problem, then, is that it is a breaking change for sites, so whether or not it is an improvement might be subjective from the perspective of a site owner. An argument in favour of landing this change even though it _is_ a change from what's on `trunk` is that it looks like the change only affects sites that already opt-in to using `blockGap` (either directly or via appearance tools). For example, if I switch `appearanceTools` to false or `blockGap` to `null`, then I can get back the extra space above the paragraph:

image

On balance, I think the changes in this PR are largely positive (good default constrained layout setting, fixed behaviour of alignment matrix for sites that use blockGap), so I'd be in favour of going with this. Would it be worth gathering more feedback from folks / a design review, to confirm the approach? From my perspective it's looking good from both a functionality and code perspective.

Great work so far! 🎉

**Comment by andrewserong** [comment]
Just one more thought, since we're opting in to the layout support, should we add support for `spacing.blockGap` at the same time, so that the spacing between blocks can be adjusted at the block level? Looks like it should work pretty well out of the box:

**Comment by ntsekouras** [comment]
>I'd be inclined to consider this an improvement instead of a regression 😅 because, although technically there is a change, that change makes spacing inside Cover consistent with the rest of the site, e.g. post content already has those spacing rules applied

I agree with the part that this is an improvement, but unfortunately is also a regression 😅   

We can get some more feedback here.. My opinion would be that we can land this(from the style changes perspective), but probably not in 6.2 at this point.

**Comment by tellthemachines** [comment]
Thanks for the extensive testing @andrewserong !

> should we add support for spacing.blockGap at the same time, so that the spacing between blocks can be adjusted at the block level?

Yes, great idea! I'll give it a go.

> My opinion would be that we can land this(from the style changes perspective), but probably not in 6.2 at this point.

Oh, definitely not in 6.2 now that we're in Beta. Let's merge it in the plugin and see how it goes for a while first!

Also cc. @WordPress/gutenberg-design for feedback, and some themes folks perhaps? @carolinan @jffng @mikachan

**Comment by jameskoster** [comment]
Seems to be working as advertised. It wasn't clear to me why I can set a wide width on children like images but not paragraphs, but that seems to be a separate thing as it occurs with Group blocks too.

Other than that, the only thing that tripped me up was this variation button:

Screenshot 2023-02-09 at 10 23 35

I don't think we need it?

**Review by mikachan — COMMENTED** [review]
This is testing well for me!

I agree that using the layout margins/spacing is an improvement, as this improves the Cover block positioning controls. But I also understand the change in spacing is a regression for existing users.

I also like the idea of supporting `blockGap` here.

I'm in favour of including this in the plugin for now and gathering more feedback that way.

**Comment by paaljoachim** [comment]
This is probably a side comment but...
Layout: "The Inner blocks use content width" setting still confuses me. I have had a hard time explaining what it actually does. The description: "Nested blocks use content width with options for full and wide widths." I get a bit tripped up by the description when two widths are used in the same sentence.

To me this is one setting that needs to be re-thought. I see the result when toggling it on in the Group block when I change the Content and Wide fields but I am still not sure how it in general affects the layout of other elements outside of it.

I am hesitant in adding the above setting to the Cover block as it just feels too vague for me and difficult to really understand. When I initially began looking at this PR I thought it was related to the "Change content position". I noticed I was wrong it was more to do with the width of the content inside the Cover block then the "Change content position" setting. 

Having a clearly defined "Change content position" would help a lot. Perhaps finding a way to simplify the "inner blocks use content width" toggle setting would help a lot as well. Finding a way that with a quick glance this specific setting becomes easier to understand.

**Comment by tellthemachines** [comment]
Thanks for the feedback everyone!

> Seems to be working as advertised. It wasn't clear to me why I can set a wide width on children like images but not paragraphs, but that seems to be a separate thing as it occurs with Group blocks too.

Yeah, I think the logic is that blocks considered "inline" like paragraphs don't have block alignment settings. It's always been like that, though perhaps it should be revisited at some point!

> Other than that, the only thing that tripped me up was this variation button:

Ah yes, that was a bug 😅 
Should be fixed now.

Regarding the addition of block spacing controls, it proved more complicated than expected because spacing styles output wasn't taking the multi-wrapper markup structure of Cover block into account. Because I had to jump through a few hoops to enable it, I decided to do so in a separate PR:  #47952. (It's currently pointing to this one so it can be properly tested with Cover block)

File: packages/block-library/src/cover/block.json
Comment (tellthemachines) [reply packages/block-library/src/cover/block.json:114]: I've looked into this a bit and I think this approach is the most consistent with the API we already have in place.
We already have the following properties:

* “allowSwitching": False by default and applicable to all types
* allowEditing": True by default and applicable to all types
* "allowInheriting": True by default and specific to flow/constrained
* "allowSizingOnChildren": False by default and specific to flex
* "allowVerticalAlignment": True by default and specific to flex
* “allowOrientation”: True by default and specific to flex

"allowJustification" (true by default) follows this pattern, and we don't even have to make it specific to "constrained", as "flex" has justification controls too. The same property could be applicable to both types.

Comment (ntsekouras) [reply packages/block-library/src/cover/block.json:114]: Thanks for the details here. I guess we can't block this PR, taking into account the above, even if it's not ideal that we have so many flags that are applied to different layouts.

**Comment by richtabor** [comment]
Nice, Something to consider: while layout is toggled off (fullwidth inner blocks), then the matrix position control is a bit confusing. You can only have top, middle, bottom positioning. 

https://user-images.githubusercontent.com/1813435/223216495-56941f07-b2af-49fc-828c-be3fb55564a1.mp4

File: packages/block-library/src/cover/index.js
Comment (richtabor) [inline packages/block-library/src/cover/index.js:16]: Do we need variations? I don't think so, yea?

**Comment by tellthemachines** [comment]
> Nice, Something to consider: while layout is toggled off (fullwidth inner blocks), then the matrix position control is a bit confusing. You can only have top, middle, bottom positioning.

Good point! Might be good to get some design input on that cc. @WordPress/gutenberg-design

File: packages/block-library/src/cover/index.js
Comment (tellthemachines) [reply packages/block-library/src/cover/index.js:16]: We do need them unfortunately. The variation here is used as a workaround to apply constrained layout to all newly added blocks, without forcing it on existing blocks, which could potentially break a lot of websites.

(Making new blocks constrained is meant to match the behaviour of Group blocks, and also mitigate the experience for hybrid themes that suddenly lose all their legacy layout styles when they add a theme.json)

**Comment by tellthemachines** [comment]
>> Nice, Something to consider: while layout is toggled off (fullwidth inner blocks), then the matrix position control is a bit confusing. You can only have top, middle, bottom positioning.

> Good point! Might be good to get some design input on that cc. https://github.com/orgs/WordPress/teams/gutenberg-design

Looking at this a bit closer, the behaviour with layout toggled off matches what we have in trunk, as currently the children of Cover always take up its full width. Likewise, on this branch, with layout off it's still possible to move the content around _as long as it doesn't take up the full block width_. For example:

withoutlayout
Trunk

withlayout
This branch

So it should be fine to keep this behaviour as is!

**Comment by richtabor** [comment]
> Looking at this a bit closer, the behaviour with layout toggled off matches what we have in trunk, as currently the children of Cover always take up its full width. 

Yea, good point.

**Review by richtabor — APPROVED** [review]
LGTM ✅

**Review by andrewserong — APPROVED** [review]
This looks like a good place to land it to me, too! Nice work 🎉

**Pull request merged**
